### ISSUE TITLE
About HTMLParser encoding

### ISSUE THREAD

**Issue by asyncins** [lestrrat-go/libxml2#76]
1、The html charset such as`utf-8/gbk/gb2312`, if html charset is `gbk` or `gb2312`, terminal show this: `encoding error : input conversion failed due to input error, bytes 0x20 0xE9 0x95 0xBF`.

I try use `doc, err := libxml2.ParseHTML(content, parser.HTMLParseIgnoreEnc)`, but the results were even worse.

What do I need to do to make the program work, to parse the html where charset is `gbk` or `gb2312`???

---

2、After reading libxml2 source code, i see this:
```
# this code from ParseHTMLString(content string, options ...parser.HTMLOption) (types.Document, error) 
docptr, err := clib.HTMLReadDoc(content, "", "", int(option))
```

Why not set `encoding="utf8"`, but set `encoding=""`?

**Comment by lestrrat** [comment]
Without a sample to work on, I can't test anything.

Did you try running your input through `xmllint` from the C libxml2? This is a wrapper around the C library, if the C library can handle it and we can't, it's probably our error, but otherwise it's your environment

**Comment by lestrrat** [comment]
Oh, you're parsing HTML... I haven't really intended this to work with HTML, so I haven't really tested this fully. I guess #70 is related.

**Comment by asyncins** [comment]
example url `http://www.changjiangtimes.com`

---

Now I've copied a copy of the code and changed it，new code is:

```
func ParseHTMLString(content string, options ...parser.HTMLOption) (types.Document, error) {
	var option parser.HTMLOption
	if len(options) > 0 {
		option = options[0]
	} else {
		option = parser.DefaultHTMLOptions
	}
	docptr, err := clib.HTMLReadDoc(content, "", "utf-8", int(option))
	if err != nil {
		return nil, errors.Wrap(err, "failed to read document")
	}

	if docptr == 0 {
		return nil, errors.Wrap(clib.ErrInvalidDocument, "failed to get valid document pointer")
	}
	return dom.WrapDocument(docptr), nil
}
```

Just add `utf-8`, the effect is good, program work

**Comment by lestrrat** [comment]
> Just add utf-8, the effect is good, program work

Right, for you. But if other people need any other value, then I need to fix it again. That's why I rejected the initial fix for #70.

Anyways, I guess I had a separate fix for #70 which I just didn't have the time to work on in my local environment -- I was apparently not happy with the second fix from #70 either.
So hopefully I will get around to working on it soon.

**Comment by lestrrat** [comment]
@asyncins Okay, so I went out and made some changes so that you could pass the encoding as an option to `ParseHTML*` methods, but I still had problems parsing the data from the URL you gave me. (in branch [topic/gh-70](https://github.com/lestrrat-go/libxml2/tree/topic/gh-70))

Turns out `xmllint` itself showed numerous errors, and I couldn't quite tell if the encoding was properly handled.

![image](https://user-images.githubusercontent.com/49281/112744670-1c48a780-8fdd-11eb-8177-32fa9f6da5c3.png)
![image](https://user-images.githubusercontent.com/49281/112744678-32566800-8fdd-11eb-8473-00205578f74d.png)

At the moment I can't tell if I'm doing something wrong, or libxml2 just can't handle the original page. I mean, I don't know what the correct answer should look like.

Please provide a *runnable*, *minimal* test case if you want this to be fixed.

**Comment by lestrrat** [comment]
Also, you should know that in general libxml2 DOES NOT automatically fix broken HTML for you. You usually need to do a fair amount of preprocessing for things to actually work.
